File: src/main/java/com/bajdcc/LALR1/interpret/module/ModuleMath.java
package com.bajdcc.LALR1.interpret.module;

import com.bajdcc.LALR1.grammar.Grammar;
import com.bajdcc.LALR1.grammar.runtime.*;
import com.bajdcc.util.ResourceLoader;

import java.math.BigDecimal;
import java.util.List;
import java.util.Random;

/**
 * 【模块】数学模块
 *
 * @author bajdcc
 */
public class ModuleMath implements IInterpreterModule {

	private static ModuleMath instance = new ModuleMath();
	private RuntimeCodePage runtimeCodePage;

	public static ModuleMath getInstance() {
		return instance;
	}

	private final static long HUNDRED = 100L;
	final static private Random rand = new Random();

	@Override
	public String getModuleName() {
		return "sys.math";
	}

	@Override
	public String getModuleCode() {
		return ResourceLoader.load(getClass());
	}

	@Override
	public RuntimeCodePage getCodePage() throws Exception {
		if (runtimeCodePage != null)
			return runtimeCodePage;

		String base = ResourceLoader.load(getClass());

		Grammar grammar = new Grammar(base);
		RuntimeCodePage page = grammar.getCodePage();
		IRuntimeDebugInfo info = page.getInfo();
		info.addExternalValue("g_PI", () -> new RuntimeObject(Math.PI));
		info.addExternalValue("g_PI_2", () -> new RuntimeObject(Math.PI * 2.0));
		info.addExternalValue("g_E", () -> new RuntimeObject(Math.E));
		info.addExternalValue("g_random", () -> new RuntimeObject(rand.nextDouble()));
		buildUnaryFunc(info);

		return runtimeCodePage = page;
	}

	private void buildUnaryFunc(IRuntimeDebugInfo info) {
		info.addExternalFunc("g_sqrt", new ModuleMathUnaryFunc("开方",
				ModuleMathUnaryFunc.ModuleMathUnaryFuncType.kSqrt));
		info.addExternalFunc("g_cos", new ModuleMathUnaryFunc("余弦",
				ModuleMathUnaryFunc.ModuleMathUnaryFuncType.kCos));
		info.addExternalFunc("g_sin", new ModuleMathUnaryFunc("正弦",
				ModuleMathUnaryFunc.ModuleMathUnaryFuncType.kSin));
		info.addExternalFunc("g_floor", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "四舍五入";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[]{RuntimeObjectType.kReal, RuntimeObjectType.kInt};
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) {
				double d = args.get(0).getDouble();
				BigDecimal decimal = BigDecimal.valueOf(d);
				long n = args.get(1).getLong();
				return new RuntimeObject(decimal.setScale((int) n, BigDecimal.ROUND_HALF_UP).doubleValue());
			}
		});
		info.addExternalFunc("g_atan2", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "atan(y, x)";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[]{RuntimeObjectType.kReal, RuntimeObjectType.kReal};
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) {
				double y = args.get(0).getDouble();
				double x = args.get(1).getDouble();
				return new RuntimeObject(Math.atan2(y, x));
			}
		});
		info.addExternalFunc("g_random_int", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "随机数";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[]{RuntimeObjectType.kInt};
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) {
				return new RuntimeObject((long) (rand.nextInt(args.get(0).getInt())));
			}
		});
	}
}

File: src/main/java/com/bajdcc/LALR1/interpret/module/ModuleMath.kt
package com.bajdcc.LALR1.interpret.module

import com.bajdcc.LALR1.grammar.Grammar
import com.bajdcc.LALR1.grammar.runtime.*
import com.bajdcc.util.ResourceLoader
import java.math.BigDecimal
import java.util.*

/**
 * 【模块】数学模块
 *
 * @author bajdcc
 */
class ModuleMath : IInterpreterModule {
    private var runtimeCodePage: RuntimeCodePage? = null

    override val moduleName: String
        get() = "sys.math"

    override val moduleCode: String
        get() = ResourceLoader.load(javaClass)

    override val codePage: RuntimeCodePage
        @Throws(Exception::class)
        get() {
            if (runtimeCodePage != null)
                return runtimeCodePage!!

            val base = ResourceLoader.load(javaClass)

            val grammar = Grammar(base)
            val page = grammar.codePage
            val info = page.info
            info.addExternalValue("g_PI", RuntimeDebugValue { RuntimeObject(Math.PI) })
            info.addExternalValue("g_PI_2", RuntimeDebugValue { RuntimeObject(Math.PI * 2.0) })
            info.addExternalValue("g_E", RuntimeDebugValue { RuntimeObject(Math.E) })
            info.addExternalValue("g_random", RuntimeDebugValue { RuntimeObject(rand.nextDouble()) })
            buildUnaryFunc(info)

            runtimeCodePage = page
            return page
        }

    private fun buildUnaryFunc(info: IRuntimeDebugInfo) {
        info.addExternalFunc("g_sqrt",
                RuntimeDebugExec("开方", arrayOf(RuntimeObjectType.kReal, RuntimeObjectType.kInt))
                { args: List<RuntimeObject>, status: IRuntimeStatus -> RuntimeObject(Math.sqrt(args[0].double)) })
        info.addExternalFunc("g_cos",
                RuntimeDebugExec("余弦", arrayOf(RuntimeObjectType.kReal, RuntimeObjectType.kInt))
                { args: List<RuntimeObject>, status: IRuntimeStatus -> RuntimeObject(Math.cos(args[0].double)) })
        info.addExternalFunc("g_sin",
                RuntimeDebugExec("正弦", arrayOf(RuntimeObjectType.kReal, RuntimeObjectType.kInt))
                { args: List<RuntimeObject>, status: IRuntimeStatus -> RuntimeObject(Math.sin(args[0].double)) })
        info.addExternalFunc("g_floor",
                RuntimeDebugExec("四舍五入", arrayOf(RuntimeObjectType.kReal, RuntimeObjectType.kInt))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val d = args[0].double
                        val decimal = BigDecimal.valueOf(d)
                        val n = args[1].long
                        RuntimeObject(decimal.setScale(n.toInt(), BigDecimal.ROUND_HALF_UP).toDouble())
                    }
                })
        info.addExternalFunc("g_atan2",
                RuntimeDebugExec("atan(y, x)", arrayOf(RuntimeObjectType.kReal, RuntimeObjectType.kReal))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val y = args[0].double
                        val x = args[1].double
                        RuntimeObject(Math.atan2(y, x))
                    }
                })
        info.addExternalFunc("g_random_int",
                RuntimeDebugExec("随机数", arrayOf(RuntimeObjectType.kInt))
                { args: List<RuntimeObject>, status: IRuntimeStatus -> RuntimeObject(rand.nextInt(args[0].int).toLong()) })
    }

    companion object {

        val instance = ModuleMath()

        private val rand = Random()
    }
}
